Revamp of networking

blork opened the pull request:
This is a pretty large change to the API, though the internals remain mostly unchanged.
This request mostly tries to address problems with how you structure and create requests, which was starting to prove unwieldy with the `Requestable` protocol. There is now a concept of an `API` which you use to create your operations and requests.

Commit:
Remove unused parameter

Commit:
Replace DecodeOperation with MapOperation

Commit:
Add encode operations

Commit:
Fix thread safety of mock session

Commit:
Update request helpers, add operations for performing multiple requests

Commit:
Rename group operations to multiple

Commit:
Better barrier protection

blork committed:
Try using resources approach from objc.io (WIP)

blork committed:
Fix tests

blork committed:
Add test for multiple upload

blork committed:
Deal with Void responses

Commit:
Update example using “Resources”

Commit:
Rename protocol to “APIProtocol”

blork committed:
Allow customisation of URL components

blork committed:
Add convenience for discarding response information

Commit:
Add comments, arrange code

Commit:
Merge branch 'feature/resources' of github.com:3squared/PeakNetwork into feature/resources

# Conflicts:
#	PeakNetwork/Core/Resources.swift

Commit:
Make resource methods extensions of API

Commit:
Add tests for endpoint

Commit:
Add misplaced tests

Commit:
Split up files, tests

Commit:
Enhance mockResponse with optional delay and block

Commit:
Add missing method parameter

Commit:
Allow constructiion of response form outside the framework

Commit:
Allow Outcome type to be properly matched

Commit:
Merge branch 'develop' into feature/resources

# Conflicts:
#	Examples/PeakNetworkExamples/WebService/WebService+Requests.swift
#	Examples/PeakNetworkExamples/WebService/WebService.swift
#	Examples/Podfile.lock
#	PeakNetwork.xcodeproj/project.pbxproj
#	PeakNetwork/Core/DecodeOperation.swift
#	PeakNetwork/Core/ImageController.swift
#	PeakNetwork/Core/NetworkOperation.swift
#	PeakNetwork/Core/Session.swift
#	Podfile.lock
#	Unit Tests/Core/NetworkTests.swift

Commit:
Update example project

Commit:
Add data task progress to NetworkOperation

Commit:
Merge branch 'develop' into feature/resources

# Conflicts:
#	PeakNetwork/Core/DecodeOperation.swift
#	PeakNetwork/Core/NetworkOperation.swift
#	PeakNetwork/Core/Session.swift

Commit:
Allow changes to a Resource after initialisation

djyates commented on PeakNetwork/Core/ImageController.swift:
```diff
@@ -60,42 +60,42 @@ public class ImageController {
         
         // Maybe we already have the image in the cache
-        let url = requestable.request.url! as NSURL
+        let url = resource.request.url! as NSURL
```
Why are we using NSURL?

blork replied:
`NSMapTable` needs it to be an `NSURL` - some old Objective-C nonsense.

djyates replied:
Bleugh

djyates commented on PeakNetwork/Core/NetworkOperation.swift:
```diff
@@ -66,34 +80,139 @@ open class NetworkOperation: RetryingOperation<NetworkResponse>, ConsumesResult
+            let operations: [NetworkOperation<O>] = resources.map { body, resource in
+                group.enter()
+
+                
```
Too many spaces

SamTTaylor commented on PeakNetwork/Core/NetworkOperation.swift:
```diff
@@ -66,34 +80,139 @@ open class NetworkOperation: RetryingOperation<NetworkResponse>, ConsumesResult
+                        Response(data: data, urlResponse: httpResponse, parsed: try resource.parse(data))
                     }
+                } else {
+                    strongSelf.output = .failure(ServerError.error(code: httpResponse.statusCodeValue, data: data, response: httpResponse))
```
Can we be certain this is a server error in this case?

BendyWalker commented on PeakNetwork/Core/API.swift:
```diff
@@ -0,0 +1,148 @@
+public protocol API {
+    
+    /// The base URL of the API. Paths will be appended to this.
+    var baseURL: String { get }
```
Shouldn't it be possible to set baseURL as either a `String` or `URL`?

blork replied:
You could, but would that be useful? It will be turned into a URL pretty quickly so you'd get a quick crash if it was invalid.

BendyWalker replied:
Realistically it doesn't matter. But just from an API standpoint, it feels like it should offer that flexibility. Or maybe it should just take a `URL`, and it's up to the user to set a valid one.

blork replied:
I have caved on this as I did with `URLQueryItems`.

BendyWalker commented on PeakNetwork/Core/API.swift:
```diff
@@ -0,0 +1,148 @@
+    var session: Session { get }
+    
+    /// Common query items that will be added to every request.
+    var commonQuery: [String: String] { get }
```
1. Is the 'common' prefix necessary here?
2. Since it's a dictionary, shouldn't it be named `queries`? If so, make sure to rename all instances of `query` in other functions.

blork replied:
They are "common" to every request. Maybe shared? I think they need something to make it clear they will be used on every request.

It represents the "query" part of the URL. It gets converted into an array of `QueryItems`, so maybe `commonQueryItems` would be more appropriate.

BendyWalker replied:
Since you're setting them on the `API`, isn't that enough context to inform they will be used on all requests?

blork replied:
I mean, I don't call it `commonSession` so I suppose that's fair.

BendyWalker commented on PeakNetwork/Core/NetworkOperation.swift:
```diff
@@ -13,27 +13,41 @@ import AppKit
 import PeakOperation
 
-public typealias NetworkResponse = (data: Data?, urlResponse: HTTPURLResponse)
+public struct Response<O> {
```
Can we make `O` a bit more descriptive?

BendyWalker commented on PeakNetwork/Core/NetworkOperation.swift:
```diff
@@ -66,34 +80,139 @@ open class NetworkOperation: RetryingOperation<NetworkResponse>, ConsumesResult
+}
+
+/// Perform a series of network requests on an internal queue and aggregate the results.
+open class MultipleResourceNetworkOperation<E, O>: ConcurrentOperation, ConsumesResult, ProducesResult {
```
I understand the utility of this operation, but feel it may be too specific or niche to include in the framework. Is the Peak Framework not meant to provide a set of building blocks?

blork replied:
This is true. Where do we keep it, then? Just as a snippet somewhere?

BendyWalker replied:
Since it was originally written for a specific internal problem, maybe we just stick it in our private 'catch-all' repo.

BendyWalker reviewed (changes requested):
Overall I like the look of these changes! There's a lot more modularity which I appreciate.

Commit:
Rename “query” to “queryItems”

Name change makes it clearer which part of the URL it represents and that it is an array.

Commit:
Changed generic parameter to full word

Better matches other uses, makes it clearer what the type is used for.

Commit:
Remove MultipleResourceNetworkOperation

djyates commented on PeakNetwork/Core/API.swift:
```diff
@@ -0,0 +1,148 @@
+
+
+/// A type that represents a JSON web API.
+public protocol JSONAPI: API {
```
Could we put `JSONAPI` in its own file?

blork replied:
no

blork replied:
@djyates How do you feel about the names of these protocols? API and JSONAPI? I'm not a huge fan.

BendyWalker replied:
I don't like API. Although it _is_ descriptive, I feel API is too broad of a term. Maybe prepend it with `Web`? And then you could have `JSONWebAPI`?

blork replied:
I like that.

djyates commented on PeakNetwork/Core/Endpoint.swift:
```diff
@@ -0,0 +1,45 @@
+public struct Endpoint {
+    let baseURL: String
+    let path: String
+    let queryItems: [String: String]
```
This should be `[String: String?]` because `URLQueryItem` can take nil for its value. This should perhaps be made in to a `typealias`

blork replied:
I have replaced `let queryItems: [String: String]` with `let queryItems: [URLQueryItem]` and made the extension that converts a dict to `[URLQueryItem]` public.

djyates commented on PeakNetwork/Core/API.swift:
```diff
@@ -0,0 +1,148 @@
+    var session: Session { get }
+    
+    /// Common query items that will be added to every request.
+    var commonQueryItems: [String: String] { get }
```
This should be `[String: String?]` because `URLQueryItem` can take nil for its value.

djyates commented on PeakNetwork/Core/Endpoint.swift:
```diff
@@ -0,0 +1,45 @@
+    }
+}
+
+extension Dictionary where Key == String, Value == String {
```
`Value == String?`

djyates commented on PeakNetwork/Core/Session.swift:
```diff
@@ -19,11 +19,12 @@ extension URLSession: Session { }
 
 /// This describes a mocked response.
 public struct MockResponse {
```
Could we move this in to a separate file?

djyates reviewed (commented):
I prefer this approach, but need to find a solution to the `[String: String?]` issue for building `URLQueryItem`s.

Commit:
Replace Dictionary with URLQueryItem array

It was noted that query items can actually have a nil value, and changing the Dictionary to [String: String?] had weird implications when when overriding the common query items. I hate URLQueryItem, but it’s easy enough to use.

Commit:
Move MockSession into a separate file

Commit:
Move JSONAPI into its own file

Commit:
Sort files

Commit:
Update example

Commit:
Change API to use URL, be strict instead of lenient with formats

Previously you could put the slashes anywhere you wanted. Let’s be more opinionated.

Commit:
Remove “common” prefix

They are obviously shared, and the other fields weren’t prefixed.

Commit:
Rename API to WebAPI

API is just too generic for a class name.

Commit:
Added missing files to macOS target

Commit:
Update example

Commit:
Make endpoint constructor public

Commit:
Remove queryItem merge logic

It is a valid URL to have multiple queryItems with the same name. This change allows duplicate query items, which means you can no longer override a shared one. If you need to do that, it’s not shared! Remove it!

Commit:
Add `passesBody` methid

This can make it a little nicer to pass the restult of a network operation into an importer, or something like that. It removes the Response wrapper, and passes just the request Body object.

Commit:
Rename unwrapped and convert to a function

Commit:
Change position of “method” parameter to resource

Commit:
Add convenience completion block for body-only

Commit:
Remove outdated README sections
